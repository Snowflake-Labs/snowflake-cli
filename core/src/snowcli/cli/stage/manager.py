from __future__ import annotations

from pathlib import Path

from snowcli.cli.common.sql_execution import SqlExecutionMixin


class StageManager(SqlExecutionMixin):
    @staticmethod
    def get_standard_stage_name(name: str) -> str:
        # Handle embedded stages
        if name.startswith("snow://"):
            return name

        return f"@{name}"

    def list(self, stage_name: str):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"ls {stage_name}")

    def get(self, stage_name: str, dest_path: Path):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"get {stage_name} file://{dest_path}/")

    def put(self, local_path: str, stage_name: str, parallel: int, overwrite: bool):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(
            f"put file://{local_path} {stage_name} "
            f"auto_compress=false parallel={parallel} overwrite={overwrite}"
        )

    def remove(self, stage_name: str, path: str):
        stage_name = self.get_standard_stage_name(stage_name)
        path = path if path.startswith("/") else "/" + path
        return self._execute_query(f"remove {stage_name}{path}")

    def show(self):
        return self._execute_query("show stages")

    def create(self, stage_name: str):
        return self._execute_query(f"create stage if not exists {stage_name}")

    def drop(self, stage_name: str):
        return self._execute_query(f"drop stage {stage_name}")
