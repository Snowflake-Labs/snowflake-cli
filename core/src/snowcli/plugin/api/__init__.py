from dataclasses import dataclass
from typing import List

import pluggy
from typer import Typer

plugin_hook_spec = pluggy.HookspecMarker("snowcli.plugin")
plugin_hook_impl = pluggy.HookimplMarker("snowcli.plugin")


@dataclass
class PluginCommandGroupPath:
    path_segments: List[str]


@dataclass
class PluginCommandGroupSpec:
    path: PluginCommandGroupPath
    command_group: Typer
