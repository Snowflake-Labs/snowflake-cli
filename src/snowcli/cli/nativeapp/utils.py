from dataclasses import dataclass
from sys import stdin, stdout
from typing import Callable, List, Optional

import typer
from snowflake.connector.cursor import DictCursor


def _rows_generator(cursor: DictCursor, predicate: Callable[[dict], bool]):
    return (row for row in cursor.fetchall() if predicate(row))


def find_all_rows(cursor: DictCursor, predicate: Callable[[dict], bool]) -> List[dict]:
    return list(_rows_generator(cursor, predicate))


def find_first_row(
    cursor: DictCursor, predicate: Callable[[dict], bool]
) -> Optional[dict]:
    """Returns the first row that matches the predicate, or None."""
    return next(_rows_generator(cursor, predicate), None)


def needs_confirmation(needs_confirm: bool, auto_yes: bool) -> bool:
    return needs_confirm and not auto_yes


class UserConfirmationPolicy:
    def __init__(self, is_force, is_interactive_mode, confirm_with_user):
        self.is_force = is_force
        self.is_interactive_mode = is_interactive_mode and not self.is_force
        self.confirm_with_user = confirm_with_user

    def requires_user_confirmation(self):
        return not self.is_force


@dataclass
class Prompts:
    give_info_and_ask: str
    do_not_continue_message: str
    error_message: str


def ask_user_confirmation(
    user_confirmation_policy: UserConfirmationPolicy, prompts: Prompts
):
    if user_confirmation_policy.requires_user_confirmation():
        if user_confirmation_policy.is_interactive_mode:
            should_continue = user_confirmation_policy.confirm_with_user(
                prompts.give_info_and_ask
            )
            if not should_continue:
                print(prompts.do_not_continue_message)
                raise typer.Exit()
        else:
            print(prompts.error_message)
            raise typer.Exit(1)
