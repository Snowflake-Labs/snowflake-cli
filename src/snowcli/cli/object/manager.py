from __future__ import annotations

from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowflake.connector.cursor import SnowflakeCursor


class ObjectManager(SqlExecutionMixin):
    def show(self, object_type: str, like: str) -> SnowflakeCursor:
        return self._execute_query(f"show {object_type} like '{like}'")

    def drop(self, object_type, name: str) -> SnowflakeCursor:
        return self._execute_query(f"drop {object_type} {name}")

    def describe(self, object_type: str, name: str):
        return self._execute_query(f"describe {object_type} {name}")
