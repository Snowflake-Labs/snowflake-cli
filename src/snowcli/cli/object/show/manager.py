from __future__ import annotations

from snowflake.connector.cursor import SnowflakeCursor

from snowcli.cli.common.sql_execution import SqlExecutionMixin


class ShowManager(SqlExecutionMixin):
    def show(self) -> SnowflakeCursor:
        return self._execute_query("show warehouses")
