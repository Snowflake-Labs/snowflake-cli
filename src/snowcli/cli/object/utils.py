from typing import Dict

from snowcli.cli.constants import ObjectType


def get_plural_name(valid_sf_name: str):
    exceptions: Dict[str, str] = {}
    return exceptions.get(valid_sf_name, f"{valid_sf_name}s")
