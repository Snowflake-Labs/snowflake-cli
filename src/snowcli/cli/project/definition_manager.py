from __future__ import annotations

import os
from pathlib import Path
from typing import List
import functools
from typing import Optional
from snowcli.exception import MissingConfiguration, InvalidPathError
from snowcli.cli.project.definition import load_project_definition


class DefinitionManager:
    BASE_DEFINITION_FILENAME = "snowflake.yml"
    USER_DEFINITION_FILENAME = "snowflake.local.yml"

    project_root: Path
    _project_config_paths: List[Path]

    def __init__(
        self,
        project_arg: Optional[str] = None,
        environment_override: Optional[str] = None,
    ):
        if environment_override:
            self.init_without_environment_override(project_arg)
        else:
            self.init_with_environment_override(
                str(project_arg), str(environment_override)
            )

    def init_without_environment_override(
        self, project_arg: Optional[str] = None
    ) -> None:
        search_path = os.getcwd()
        if project_arg:
            search_path = os.path.abspath(project_arg)
        config_files = self._find_definition_files(Path(search_path))
        if not config_files:
            raise MissingConfiguration(
                f"Cannot find project definition (snowflake.yml). Please provide a path to the project or run this command in a valid project directory."
            )
        self._project_config_paths = config_files

    def init_with_environment_override(
        self, project_path: str, environment_override: str
    ) -> None:
        project_root = Path(project_path)
        if project_root.exists():
            self.project_root = project_root
        else:
            raise InvalidPathError(
                f"The path {project_path} does not exist. Please provide a valid path to the project or run this command in a valid project directory."
            )

        override_path = Path.joinpath(project_root, environment_override)
        if not self._base_definition_file_if_available(override_path.parent):
            raise MissingConfiguration(
                f"Cannot find project definition (snowflake.yml) at {override_path.parent}. Please provide a valid path to the project or run this command in a valid project directory."
            )
        self._project_config_paths = [
            override_path.parent / self.BASE_DEFINITION_FILENAME
        ]

        if override_path.name != self.BASE_DEFINITION_FILENAME:
            if (
                self._definition_if_available(override_path.name, override_path.parent)
                is None
            ):
                raise MissingConfiguration(
                    f"Cannot find project definition {override_path.name} at {override_path.parent}. Please provide a valid path to the project or run this command in a valid project directory."
                )
            self._project_config_paths.append(override_path)

    def _find_definition_files(self, project_path: Path) -> Optional[List[Path]]:
        """
        Recurse up the directory tree from the given search path until we find
        a directory that contains a snowflake.yml file. We'll stop if we cross
        a filesystem boundary or hit the user's HOME directory.
        """
        parent_path = project_path
        starting_mount = project_path.is_mount()
        while parent_path:
            if (
                project_path.is_mount() != starting_mount
                or parent_path.parent == parent_path
                or parent_path == Path.home()
            ):
                return None

            base_config_file_path = self._base_definition_file_if_available(parent_path)
            if base_config_file_path:
                user_config_file_path = self._user_definition_file_if_available(
                    parent_path
                )
                self.project_root = parent_path
                if user_config_file_path:
                    return [
                        base_config_file_path,
                        user_config_file_path,
                    ]
                return [base_config_file_path]

            parent_path = parent_path.parent
        return None

    def _definition_if_available(self, filename, project_path: Path) -> Optional[Path]:
        file_path = Path(project_path) / filename
        if file_path.is_file():
            return file_path
        return None

    def _base_definition_file_if_available(self, project_path: Path) -> Optional[Path]:
        return self._definition_if_available(
            self.BASE_DEFINITION_FILENAME, project_path
        )

    def _user_definition_file_if_available(self, project_path: Path) -> Optional[Path]:
        return self._definition_if_available(
            self.USER_DEFINITION_FILENAME, project_path
        )

    @functools.cached_property
    def project_definition(self) -> dict:
        return load_project_definition(self._project_config_paths)
