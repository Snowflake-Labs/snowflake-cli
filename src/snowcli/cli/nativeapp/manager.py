from __future__ import annotations
from collections import OrderedDict

import logging
from pathlib import Path
from functools import cached_property
from typing import List, Optional
from click.exceptions import ClickException
from snowcli.exception import SnowflakeSQLExecutionError

from snowflake.connector.cursor import SnowflakeCursor

from snowcli.cli.project.util import clean_identifier
from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowcli.cli.project.definition import (
    default_app_package,
    default_role,
    default_application,
)
from snowcli.output.printing import print_result
from snowcli.output.types import ObjectResult
from snowcli.cli.stage.diff import (
    DiffResult,
    stage_diff,
    sync_local_diff_with_stage,
)
from snowcli.cli.project.definition_manager import DefinitionManager
from snowcli.cli.nativeapp.artifacts import (
    build_bundle,
    translate_artifact,
    ArtifactMapping,
)

from snowflake.connector.cursor import DictCursor

SPECIAL_COMMENT = "GENERATED_BY_SNOWCLI"
COMMENT_COL = "comment"

log = logging.getLogger(__name__)


class ApplicationPackageAlreadyExistsError(ClickException):
    def __init__(self, name: str):
        super().__init__(
            f"An Application Package {name} already exists in account that may have been created without snowCLI. "
        )


class CouldNotDropObjectError(ClickException):
    """
    Could not successfully drop the required Snowflake object.
    """

    def __init__(self, message: str):
        super().__init__(message=message)


class NativeAppManager(SqlExecutionMixin):
    definition_manager: DefinitionManager

    def __init__(self, search_path: Optional[str] = None):
        super().__init__()
        self.definition_manager = DefinitionManager(search_path or "")

    @property
    def project_root(self) -> Path:
        return self.definition_manager.project_root

    @property
    def definition(self) -> dict:
        return self.definition_manager.project_definition["native_app"]

    @cached_property
    def artifacts(self) -> List[ArtifactMapping]:
        return [translate_artifact(item) for item in self.definition["artifacts"]]

    @cached_property
    def deploy_root(self) -> Path:
        return Path(self.project_root, self.definition["deploy_root"])

    def build_bundle(self) -> None:
        build_bundle(self.project_root, self.deploy_root, self.artifacts)

    def sync_deploy_root_with_stage(self, role: str, app_pkg: str):
        schema_and_stage: str = self.definition["source_stage"]

        # Does a stage already exist within the app pkg, or we need to create one?
        # Using "if not exists" should take care of either case.
        log.info("Checking if stage exists, or creating a new one if none exists.")
        with self.use_role(role):
            self._execute_query(
                f"create schema if not exists {app_pkg}.{schema_and_stage.split('.')[0]}"
            )
            self._execute_query(
                f"""
                    create stage if not exists {app_pkg}.{schema_and_stage}
                    encryption = (TYPE = 'SNOWFLAKE_SSE')"""
            )

        # Perform a diff operation and display results to the user for informational purposes
        log.info(
            f"Performing a diff between the Snowflake stage and your local deploy_root {self.deploy_root} directory."
        )
        diff: DiffResult = stage_diff(self.deploy_root, f"{app_pkg}.{schema_and_stage}")
        log.info("Listing results of diff:")
        log.info(f"New files only on your local: {','.join(diff.only_local)}")
        log.info(f"New files only on the stage: {','.join(diff.only_on_stage)}")
        log.info(
            f"Existing files modified or status unknown: {','.join(diff.different)}"
        )
        log.info(f"Existing files identical to the stage: {','.join(diff.identical)}")

        # Upload diff-ed files to app pkg stage
        sync_local_diff_with_stage(
            role=role,
            deploy_root_path=self.deploy_root,
            diff_result=diff,
            stage_path=f"{app_pkg}.{schema_and_stage}",
        )

    def app_run(self) -> SnowflakeCursor:

        project_name = clean_identifier(self.definition["name"])
        app_pkg = self.definition.get("package", {}).get(
            "name", default_app_package(project_name)
        )
        app_pkg_role = self.definition.get("package", {}).get("role", default_role())

        with self.use_role(app_pkg_role):
            show_cursor = self._execute_query(
                f"show application packages like '{app_pkg}'",
                cursor_class=DictCursor,
            )

            if show_cursor.rowcount is None:
                raise SnowflakeSQLExecutionError()
            elif show_cursor.rowcount == 0:
                # Create an app pkg, with distribution = internal to avoid triggering security scan
                log.info(f"Creating new application package {app_pkg} in account.")
                self._execute_query(
                    f"""
                    create application package {app_pkg}
                        comment = {SPECIAL_COMMENT}
                        distribution = internal
                """
                )
            else:
                # There can only be one possible pre-existing app pkg with the same name
                show_app_row = show_cursor.fetchone()
                row_comment = show_app_row[
                    COMMENT_COL
                ]  # Because we use a DictCursor, we are guaranteed a dictionary instead of indexing through a list.

                if row_comment != SPECIAL_COMMENT:
                    raise ApplicationPackageAlreadyExistsError(app_pkg)

            # Upload files from deploy root local folder to the above stage
            self.sync_deploy_root_with_stage(app_pkg_role, app_pkg)

    def drop_object(
        self,
        object_name: str,
        object_role: str,
        object_type: str,
        query_dict: dict,
    ) -> None:
        log_object_type = (
            "Application Package" if object_type == "package" else object_type
        )

        with self.use_role(object_role):
            show_obj_cursor = self._execute_query(
                f"{query_dict['show']} '{object_name}'", cursor_class=DictCursor
            )

            if show_obj_cursor.rowcount == 0:
                raise CouldNotDropObjectError(
                    f"Role {object_role} does not own any {log_object_type.lower()} with name {object_name}!"
                )
            elif show_obj_cursor.rowcount > 0:
                # There can only be one possible pre-existing app pkg with the same name
                show_obj_row = show_obj_cursor.fetchone()
                row_comment = show_obj_row[
                    COMMENT_COL
                ]  # Because we use a DictCursor, we are guaranteed a dictionary instead of indexing through a list.

                if row_comment != SPECIAL_COMMENT:
                    raise CouldNotDropObjectError(
                        f"{log_object_type} {object_name} was not created by SnowCLI. Cannot drop the {log_object_type.lower()}."
                    )
            else:  # rowcount is None
                raise SnowflakeSQLExecutionError()

            log.info(f"Dropping {log_object_type.lower()} {object_name} now.")
            self._execute_query(f"{query_dict['drop']} {object_name}")
            log.info(f"Dropped {log_object_type.lower()} {object_name} successfully.")

    def teardown(self) -> None:
        project_name = clean_identifier(self.definition["name"])

        # Drop the application first
        self.drop_object(
            object_name=self.definition.get("application", {}).get(
                "name", default_application(project_name)
            ),
            object_role=self.definition.get("application", {}).get(
                "role", default_role(project_name)
            ),
            object_type="application",
            query_dict={"show": "show applications like", "drop": "drop application"},
        )

        # Drop the application package next
        self.drop_object(
            object_name=self.definition.get("package", {}).get(
                "name", default_app_package(project_name)
            ),
            object_role=self.definition.get("package", {}).get(
                "role", default_role(project_name)
            ),
            object_type="package",
            query_dict={
                "show": "show application packages like",
                "drop": "drop application package",
            },
        )
