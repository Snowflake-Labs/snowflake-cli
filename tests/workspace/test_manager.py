# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import annotations

import os
from unittest import mock

import pytest
from snowflake.cli.api.exceptions import InvalidProjectDefinitionVersionError
from snowflake.cli.api.project.definition_manager import DefinitionManager
from snowflake.cli.plugins.workspace.manager import WorkspaceManager

from tests.testing_utils.files_and_dirs import create_named_file
from tests.workspace.utils import (
    APP_PACKAGE_ENTITY,
    MOCK_SNOWFLAKE_YML_FILE,
    MOCK_SNOWFLAKE_YML_V1_FILE,
)


def _get_ws_manager(pdf_content=MOCK_SNOWFLAKE_YML_FILE):
    current_working_directory = os.getcwd()
    create_named_file(
        file_name="snowflake.yml",
        dir_name=current_working_directory,
        contents=[pdf_content],
    )
    dm = DefinitionManager()
    return WorkspaceManager(
        project_definition=dm.project_definition,
        project_root=dm.project_root,
    )


def test_pdf_not_v2(temp_dir):
    with pytest.raises(InvalidProjectDefinitionVersionError):
        _get_ws_manager(pdf_content=MOCK_SNOWFLAKE_YML_V1_FILE)


# Test that the same entity instance is returned for the same key
def test_get_entity_is_cached(temp_dir):
    ws_manager = _get_ws_manager()
    pkg1 = ws_manager.get_entity("pkg")
    pkg2 = ws_manager.get_entity("pkg")
    app = ws_manager.get_entity("app")
    assert pkg1 is pkg2
    assert app is not pkg1


def test_get_entity_invalid_key(temp_dir):
    ws_manager = _get_ws_manager()
    with pytest.raises(ValueError, match="No such entity key"):
        ws_manager.get_entity("non_existing_key")


def test_bundle(temp_dir):
    ws_manager = _get_ws_manager()
    with mock.patch(f"{APP_PACKAGE_ENTITY}.bundle") as app_pkg_bundle_mock:
        ws_manager.bundle("pkg")
        app_pkg_bundle_mock.assert_called_once()


def test_bundle_of_invalid_entity_type(temp_dir):
    ws_manager = _get_ws_manager()
    with pytest.raises(ValueError, match="This entity type does not support bundling"):
        ws_manager.bundle("app")
