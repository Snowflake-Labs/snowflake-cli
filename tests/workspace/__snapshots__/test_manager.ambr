# serializer version: 1
# name: test_if_template_is_not_rendered_during_migration[snowpark_templated_v1]
  '''
  defaults:
    project_name: <! project_name | to_snowflake_identifier !>
    stage: <! stage | to_snowflake_identifier !>
  definition_version: '2'
  entities:
    hello_function:
      artifacts:
      - <% ctx.env.project_source %>
      external_access_integrations: []
      handler: functions.hello_function
      identifier:
        name: hello_function
      imports: []
      returns: string
      secrets: {}
      signature:
      - arg_type: string
        name: name
      stage: <! stage | to_snowflake_identifier !>
      type: function
    hello_procedure:
      artifacts:
      - <% ctx.env.project_source %>
      execute_as_caller: false
      external_access_integrations: []
      handler: procedures.hello_procedure
      identifier:
        name: hello_procedure
      imports: []
      returns: string
      secrets: {}
      signature:
      - arg_type: string
        name: name
      stage: <! stage | to_snowflake_identifier !>
      type: procedure
    test_procedure:
      artifacts:
      - <% ctx.env.project_source %>
      execute_as_caller: false
      external_access_integrations: []
      handler: procedures.test_procedure
      identifier:
        name: test_procedure
      imports: []
      returns: string
      secrets: {}
      signature: ''
      stage: <! stage | to_snowflake_identifier !>
      type: procedure
  
  '''
# ---
# name: test_if_template_is_not_rendered_during_migration[snowpark_templated_v1].1
  '''
  definition_version: "1.1"
  snowpark:
    project_name: "<! project_name | to_snowflake_identifier !>"
    stage_name: "<! stage | to_snowflake_identifier !>"
    src: <% ctx.env.project_source %>
    functions:
      - name: hello_function
        handler: "functions.hello_function"
        signature:
          - name: "name"
            type: "string"
        returns: string
    procedures:
      - name: hello_procedure
        handler: "procedures.hello_procedure"
        signature:
          - name: "name"
            type: "string"
        returns: string
      - name: test_procedure
        handler: "procedures.test_procedure"
        signature: ""
        returns: string
  
  '''
# ---
# name: test_if_template_is_not_rendered_during_migration[streamlit_templated_v1]
  '''
  definition_version: '2'
  entities:
    <! name | to_snowflake_identifier !>:
      artifacts:
      - streamlit_app.py
      - environment.yml
      - pages
      - common/hello.py
      identifier:
        name: <! name | to_snowflake_identifier !>
      main_file: streamlit_app.py
      pages_dir: pages
      query_warehouse: <! query_warehouse | to_snowflake_identifier !>
      stage: <! stage | to_snowflake_identifier !>
      title: <% ctx.env.streamlit_title %>
      type: streamlit
  
  '''
# ---
# name: test_if_template_is_not_rendered_during_migration[streamlit_templated_v1].1
  '''
  definition_version: "1.1"
  streamlit:
    name: <! name | to_snowflake_identifier !>
    stage: <! stage | to_snowflake_identifier !>
    query_warehouse: <! query_warehouse | to_snowflake_identifier !>
    main_file: streamlit_app.py
    env_file: environment.yml
    pages_dir: pages/
    title: <% ctx.env.streamlit_title %>
    additional_source_files:
      - common/hello.py
  
  '''
# ---
# name: test_migration_streamlit_v1_to_v2[migration_snowpark_V1_to_V2]
  '''
  defaults:
    project_name: my_snowpark_project
    stage: dev_deployment
  definition_version: '2'
  entities:
    func1:
      artifacts: app/
      external_access_integrations: []
      handler: app.func1_handler
      identifier:
        name: func1
      imports: []
      returns: string
      runtime: '3.10'
      secrets: {}
      signature:
      - arg_type: string
        default: default value
        name: a
      - arg_type: variant
        name: b
      stage: dev_deployment
      type: function
    procedureName:
      artifacts: app/
      execute_as_caller: false
      external_access_integrations: []
      handler: hello
      identifier:
        name: procedureName
      imports: []
      returns: string
      secrets: {}
      signature:
      - arg_type: string
        name: name
      stage: dev_deployment
      type: procedure
    test:
      artifacts: app/
      execute_as_caller: false
      external_access_integrations: []
      handler: test
      identifier:
        name: test
      imports: []
      returns: string
      runtime: '3.10'
      secrets: {}
      signature: ''
      stage: dev_deployment
      type: procedure
  
  '''
# ---
# name: test_migration_streamlit_v1_to_v2[migration_snowpark_V1_to_V2].1
  '''
  definition_version: 1
  snowpark:
    project_name: "my_snowpark_project"
    stage_name: "dev_deployment"
    src: "app/"
    functions:
      - name: func1
        handler: "app.func1_handler"
        signature:
          - name: "a"
            type: "string"
            default: "default value"
          - name: "b"
            type: "variant"
        returns: string
        runtime: 3.10
    procedures:
      - name: procedureName
        handler: "hello"
        signature:
          - name: "name"
            type: "string"
        returns: string
      - name: test
        handler: "test"
        signature: ""
        returns: string
        runtime: "3.10"
  
  '''
# ---
# name: test_migration_streamlit_v1_to_v2[migration_streamlit_V1_to_V2]
  '''
  definition_version: '2'
  entities:
    test_streamlit:
      artifacts:
      - streamlit_app.py
      - environment.yml
      - pages
      identifier:
        name: test_streamlit
      main_file: streamlit_app.py
      pages_dir: None
      query_warehouse: test_warehouse
      stage: streamlit
      title: My Fancy Streamlit
      type: streamlit
  
  '''
# ---
# name: test_migration_streamlit_v1_to_v2[migration_streamlit_V1_to_V2].1
  '''
  definition_version: 1
  streamlit:
    name: test_streamlit
    stage: streamlit
    query_warehouse: test_warehouse
    main_file: "streamlit_app.py"
    title: "My Fancy Streamlit"
  
  '''
# ---
# name: test_migration_v1_to_v2[migration_snowpark_V1_to_V2]
  '''
  defaults:
    project_name: my_snowpark_project
    stage: dev_deployment
  definition_version: '2'
  entities:
    func1:
      artifacts:
      - app
      external_access_integrations: []
      handler: app.func1_handler
      identifier:
        name: func1
      imports: []
      returns: string
      runtime: '3.10'
      secrets: {}
      signature:
      - arg_type: string
        default: default value
        name: a
      - arg_type: variant
        name: b
      stage: dev_deployment
      type: function
    procedureName:
      artifacts:
      - app
      execute_as_caller: false
      external_access_integrations: []
      handler: hello
      identifier:
        name: procedureName
      imports: []
      returns: string
      secrets: {}
      signature:
      - arg_type: string
        name: name
      stage: dev_deployment
      type: procedure
  
  '''
# ---
# name: test_migration_v1_to_v2[migration_snowpark_V1_to_V2].1
  '''
  definition_version: 1
  snowpark:
    project_name: "my_snowpark_project"
    stage_name: "dev_deployment"
    src: "app/"
    functions:
      - name: func1
        handler: "app.func1_handler"
        signature:
          - name: "a"
            type: "string"
            default: "default value"
          - name: "b"
            type: "variant"
        returns: string
        runtime: 3.10
    procedures:
      - name: procedureName
        handler: "hello"
        signature:
          - name: "name"
            type: "string"
        returns: string
  
  '''
# ---
# name: test_migration_v1_to_v2[migration_streamlit_V1_to_V2]
  '''
  definition_version: '2'
  entities:
    test_streamlit:
      artifacts:
      - streamlit_app.py
      - environment.yml
      - pages
      identifier:
        name: test_streamlit
      main_file: streamlit_app.py
      pages_dir: None
      query_warehouse: test_warehouse
      stage: streamlit
      title: My Fancy Streamlit
      type: streamlit
  
  '''
# ---
# name: test_migration_v1_to_v2[migration_streamlit_V1_to_V2].1
  '''
  definition_version: 1
  streamlit:
    name: test_streamlit
    stage: streamlit
    query_warehouse: test_warehouse
    main_file: "streamlit_app.py"
    title: "My Fancy Streamlit"
  
  '''
# ---
